xla/stream_executor/gpu/gpu_command_buffer.cc

```diff
@@ -239,9 +239,37 @@ tsl::Status GpuCommandBuffer::MemcpyDeviceToDevice(DeviceMemoryBase* dst,
   if (state_ == State::kCreate) {
     Dependencies deps = GetDependencies();
     GpuGraphNodeHandle* node = &nodes_.emplace_back();
-    return GpuDriver::GraphAddMemcpyD2DNode(
-        parent_->gpu_context(), node, graph_, absl::MakeSpan(deps),
-        AsDevicePtr(*dst), AsDevicePtr(src), size);
+    return GpuDriver::GraphAddMemcpyD2DNode(parent_->gpu_context(), node,
+                                            graph_, absl::MakeSpan(deps), AsDevicePtr(*dst),
+                                            AsDevicePtr(src), size);
+
+    return UnsupportedStateError(state_);
+  }
+}
+
+tsl::Status GpuCommandBuffer::Allocate(uint64_t size, DeviceMemoryBase* dptr) {
+  TF_RETURN_IF_ERROR(CheckNotFinalized());
+
+  // Adds a new memalloc node to the graph under construction.
+  if (state_ == State::kCreate) {
+    Dependencies deps = GetDependencies();
+    GpuGraphNodeHandle* node = &nodes_.emplace_back();
+    GpuDevicePtr gpu_dptr;
+    TF_RETURN_IF_ERROR(GpuDriver::GraphAddMemAllocNode(
+        node, graph_, absl::MakeSpan(deps), GpuDriver::MemAccessFlags::kReadWrite,
+        GpuDriver::MemLocationType::kDevice, parent_->device_ordinal(),
+        GpuDriver::MemAllocationType::kPinned, size, &gpu_dptr));
+    // For CUDA impl, VA range is reserved when adding memory allocation node.
+    CHECK(gpu_dptr) << "Graph add memory allocation node return null ptr";
+    VLOG(2) << "Setting device memory base with opaque pointer "
+            << reinterpret_cast<void*>(gpu_dptr)
+            << " device ordinal: " << parent_->device_ordinal();
+    dptr->SetOpaque(reinterpret_cast<void*>(gpu_dptr));
+    return tsl::OkStatus();
+  } else if (state_ == State::kUpdate) {
+    // memalloc node does not need to update, just step the node update idx
+    node_update_idx_++;
```

Comment:
you'll need to keep a separate vector of `se::DeviceMemoryBase` to return memory ptr allocated at `kCreate` time earlier